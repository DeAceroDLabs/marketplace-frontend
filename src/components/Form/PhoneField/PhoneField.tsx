import { OptionsField } from "forms/form.types";
import { useState } from "react";
import { useFormContext } from "react-hook-form";
import styles from "./PhoneField.module.scss";
import validator from "validator";

const PhoneField: React.FC<OptionsField> = ({
  name,
  value,
  placeholder,
  label,
  required,
  type,
  disabled,
  errorMessage = "",
}) => {
  const [error, setError] = useState("");
  const methods = useFormContext();
  const [currentValue, setCurrentValue] = useState(value);
  const validatePhone = (phoneValue: string) => {
    validator.isMobilePhone(phoneValue, "es-MX")
      ? setError("")
      : setError(errorMessage);
  };

  const noError = error !== "";
  const emptyFieldWhenRequired =
    methods.formState.errors[name] &&
    methods.formState.errors[name].type === "required";
  const valueNotChanged = currentValue === value;

  const errorStyle = noError || (emptyFieldWhenRequired && valueNotChanged) ? "input-error" : "";

  const requiredMessage = emptyFieldWhenRequired && valueNotChanged && (
    <span className={styles["error-text"]}>Este campo es requerido</span>
  );

  noError && (methods.formState.errors[name] = error);

  return (
    <div className={styles.container}>
      <label>{label}</label>
      <input
        {...methods.register(name, { value, required })}
        className={`${styles.input} ${styles[errorStyle]}`}
        value={currentValue}
        type={type}
        placeholder={placeholder}
        disabled={disabled}
        onChange={(e) => {
          let phoneValue = e.target.value;
          const noPrefix = phoneValue.slice(0, 3) !== "+52" && phoneValue.length === 10;
          phoneValue = noPrefix ? `+52${phoneValue}` : phoneValue;
          setCurrentValue(phoneValue);
          validatePhone(phoneValue);
        }}
      />
      {requiredMessage}
      {error !== "" && <div className={styles["error-text"]}> {error}</div>}
    </div>
  );
};

export default PhoneField;
