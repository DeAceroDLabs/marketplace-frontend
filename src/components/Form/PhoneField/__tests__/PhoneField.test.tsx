import { render } from "@testing-library/react";
import { Option } from "forms/form.types";
import PhoneField from "../PhoneField";
import { fireEvent, screen, waitFor } from "@testing-library/react";

jest.mock("react-hook-form", () => ({
  ...jest.requireActual("react-hook-form"),
  useFormContext: () => ({
    register: () => jest.fn(),
    handleSubmit: () => jest.fn(),
    errors: () => jest.fn(),
    formState: { errors: {} },
  }),
}));

describe("PhoneField", () => {
  const setup = (options: Option[]) => {
    return render(
      <PhoneField
        label={"Mock PhoneField"}
        required={false}
        placeholder={"+52 (000)-1111-222"}
        type={"dropdown"}
        value={"mock value"}
        name={"mock name"}
        errorMessage={"mock error message"}
        validateDomain={"mock validateDomain"}
      />
    );
  };

  it("renders PhoneField with no options without problem", () => {
    const view = setup([]);
    expect(view).toMatchSnapshot();
  });

  it("simulates change on input", async () => {
    const view = setup([]);
    const input = screen.getByPlaceholderText("+52 (000)-1111-222");
    fireEvent.change(input, {
      target: { value: "8180199403" },
    });
    await waitFor(() => {
        expect((input as HTMLInputElement).value).toBe("8180199403");
    });
  });

  it("simulates change on input", async () => {
    const view = setup([]);
    const input = screen.getByPlaceholderText("+52 (000)-1111-222");
    fireEvent.change(input, {
      target: { value: "12345678" },
    });
    await waitFor(() => {
      expect(screen.getByText("mock error message")).toBeInTheDocument();
    });
  });
  
});
