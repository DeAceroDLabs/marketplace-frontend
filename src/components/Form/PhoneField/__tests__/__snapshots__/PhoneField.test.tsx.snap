// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneField renders PhoneField with no options without problem 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <label>
          Mock PhoneField
        </label>
        <input
          class="input "
          mock="[object Object]"
          placeholder="+52 (000)-1111-222"
          type="number"
          value="mock value"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <label>
        Mock PhoneField
      </label>
      <input
        class="input "
        mock="[object Object]"
        placeholder="+52 (000)-1111-222"
        type="number"
        value="mock value"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
